### PR TITLE
[CARBONDATA-2853] Implement min/max index for streaming segment

### PR DESCRIPTION
**Pull request by QiangCai**
Implement file-level min/max index to prune stream files in driver side.

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6295/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7944/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6667/

**Comment by QiangCai** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7959/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6683/

File: core/src/main/java/org/apache/carbondata/core/datamap/StreamDataMap.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+import org.apache.carbondata.core.reader.CarbonIndexFileReader;
+import org.apache.carbondata.core.scan.filter.FilterUtil;
+import org.apache.carbondata.core.scan.filter.executer.FilterExecuter;
+import org.apache.carbondata.core.scan.filter.executer.ImplicitColumnFilterExecutor;
+import org.apache.carbondata.core.scan.filter.resolver.FilterResolverIntf;
+import org.apache.carbondata.core.util.CarbonMetadataUtil;
+import org.apache.carbondata.core.util.path.CarbonTablePath;
+import org.apache.carbondata.format.BlockIndex;
+
+public class StreamDataMap {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datamap/StreamDataMap.java]: add interface annotation

File: core/src/main/java/org/apache/carbondata/core/datamap/StreamFile.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.carbondata.core.datamap;
+
+import org.apache.carbondata.core.metadata.blocklet.index.BlockletMinMaxIndex;
+
+public class StreamFile {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datamap/StreamFile.java]: For all public class, please add interface annotation

File: core/src/main/java/org/apache/carbondata/core/util/ByteUtil.java [hunk 3]
```diff
@@ -669,4 +666,44 @@ public static int putBytes(byte[] tgtBytes, int tgtOffset, byte[] srcBytes, int
...
+  }
+
+  /**
+   * convert byte array to the value, perform XOR operation on it to recover the real value
+   */
+  public static short toXorShort(byte[] bytes, int offset, final int length) {
+    return (short)(toShort(bytes, offset, length) ^ Short.MIN_VALUE);
+  }
+
+  public static int toXorInt(byte[] bytes, int offset, final int length) {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/util/ByteUtil.java]: please add comment

File: streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java [hunk 4]
```diff
@@ -171,9 +179,57 @@ private void initializeAtFirstRow() throws IOException, InterruptedException {
       writeFileHeader();
     }
 
+    initializeStatsCollector();
+
     isFirstRow = false;
   }
 
+  private void initializeStatsCollector() {
+    // initialize
```
Comment (jackylk) [inline streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java]: please explain why the length is isNoDictionaryDimensionColumn.length

File: streaming/src/main/java/org/apache/carbondata/streaming/index/StreamFileIndex.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+package org.apache.carbondata.streaming.index;
+
+import java.io.Serializable;
+
+import org.apache.carbondata.core.metadata.blocklet.index.BlockletMinMaxIndex;
+import org.apache.carbondata.core.metadata.datatype.DataType;
+
+public class StreamFileIndex implements Serializable {
+
+  private String fileName;
```
Comment (jackylk) [inline streaming/src/main/java/org/apache/carbondata/streaming/index/StreamFileIndex.java]: please describe the content of `fileName`, whether it includes the whole path

File: core/src/main/java/org/apache/carbondata/core/datamap/DataMapStoreManager.java [hunk 6]
```diff
@@ -334,6 +334,10 @@ public TableDataMap getDataMap(CarbonTable table, DataMapSchema dataMapSchema) {
     return dataMap;
   }
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datamap/DataMapStoreManager.java]: please add a testcase to test the bigint datatype, for overflow scenario

File: core/src/main/java/org/apache/carbondata/core/datamap/StreamDataMap.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        if (hitStreamFile(streamFile)) {
+          streamFileList.add(streamFile);
+          streamFile.setMinMaxIndex(null);
+        }
+      }
+      return streamFileList;
+    }
+  }
+
+  private boolean hitStreamFile(StreamFile streamFile) {
```
Comment (xuchuanyin) [inline core/src/main/java/org/apache/carbondata/core/datamap/StreamDataMap.java]: I think change the name to 'isScanRequired' is better.

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6405/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8063/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/5/

File: core/src/main/java/org/apache/carbondata/core/datamap/StreamDataMap.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+  private boolean isScanRequire(StreamFile streamFile) {
+    // backward compatibility, old stream file without min/max index
+    if (streamFile.getMinMaxIndex() == null) {
+      return true;
+    }
+
+    byte[][] maxValue = streamFile.getMinMaxIndex().getMaxValues();
+    byte[][] minValue = streamFile.getMinMaxIndex().getMinValues();
+    BitSet bitSet;
+    if (filterExecuter instanceof ImplicitColumnFilterExecutor) {
```
Comment (xuchuanyin) [inline core/src/main/java/org/apache/carbondata/core/datamap/StreamDataMap.java]: What does this for?

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8193/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/122/

**Comment by manishgupta88** [comment]
@QiangCai ....In General I can see that you put empty lines at many places in the code. Please remove those empty lines everywhere and add some code comments for better understanding

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/173/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8244/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8255/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/184/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8284/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/213/

File: core/src/main/java/org/apache/carbondata/core/datamap/StreamDataMap.java [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+  private List<StreamFile> listAllStreamFiles(List<Segment> segments, boolean withMinMax)
+      throws IOException {
+    List<StreamFile> streamFileList = new ArrayList<>();
+    for (Segment segment : segments) {
+      String segmentDir = CarbonTablePath.getSegmentPath(
+          carbonTable.getAbsoluteTableIdentifier().getTablePath(), segment.getSegmentNo());
+      FileFactory.FileType fileType = FileFactory.getFileType(segmentDir);
+      if (FileFactory.isFileExist(segmentDir, fileType)) {
+        SegmentIndexFileStore segmentIndexFileStore = new SegmentIndexFileStore();
+        segmentIndexFileStore.readAllIIndexOfSegment(segmentDir);
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/datamap/StreamDataMap.java]: It involves the listing of a folder and gets the index file. Its a costly operation for s3 file system as you are not even caching it. So it is better fix the carbonindex file name in case of streaming and read it directly instead of listing.

File: core/src/main/java/org/apache/carbondata/core/datamap/StreamFile.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    this.segmentNo = segmentNo;
+    this.filePath = filePath;
+    this.fileSize = fileSize;
+  }
+
+  public String getSegmentNo() {
+    return segmentNo;
+  }
+
+  public void setSegmentNo(String segmentNo) {
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/datamap/StreamFile.java]: Please remove unused setters

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java [hunk 11]
```diff
@@ -339,63 +338,62 @@ public void refreshSegmentCacheIfRequired(JobContext job, CarbonTable carbonTabl
...
+          }
+        }
+      }
+      StreamDataMap streamDataMap =
+          DataMapStoreManager.getInstance().getStreamDataMap(carbonTable);
+      streamDataMap.init(filterResolverIntf);
+      List<StreamFile> streamFiles = streamDataMap.prune(streamSegments);
+
+      for (StreamFile streamFile : streamFiles) {
+        if (FileFactory.isFileExist(streamFile.getFilePath())) {
```
Comment (ravipesala) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java]: Why do you need to do file exists check here? Please remove if not needed, will be costly for s3

File: streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java [hunk 12]
```diff
@@ -212,9 +213,13 @@ private void initializeAtFirstRow() throws IOException, InterruptedException {
             byte[] col = (byte[]) columnValue;
             output.writeShort(col.length);
             output.writeBytes(col);
+            output.dimStatsCollectors[dimCount].update(col);
           } else {
             output.writeInt((int) columnValue);
+            output.dimStatsCollectors[dimCount].update(ByteUtil.toBytes((int) columnValue));
```
[the remark this answers is not in the document]
Comment (QiangCai) [reply streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java]: update(int ) isn't be implemented in KeyPageStatsCollector

File: core/src/main/java/org/apache/carbondata/core/datamap/StreamDataMap.java [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+import org.apache.carbondata.core.reader.CarbonIndexFileReader;
+import org.apache.carbondata.core.scan.filter.FilterUtil;
+import org.apache.carbondata.core.scan.filter.executer.FilterExecuter;
+import org.apache.carbondata.core.scan.filter.resolver.FilterResolverIntf;
+import org.apache.carbondata.core.util.CarbonMetadataUtil;
+import org.apache.carbondata.core.util.path.CarbonTablePath;
+import org.apache.carbondata.format.BlockIndex;
+
+@InterfaceAudience.Internal
+public class StreamDataMap {
```
[the remark this answers is not in the document]
Comment (QiangCai) [reply core/src/main/java/org/apache/carbondata/core/datamap/StreamDataMap.java]: fixed

File: core/src/main/java/org/apache/carbondata/core/datamap/DataMapStoreManager.java [hunk 14]
```diff
@@ -360,6 +360,10 @@ private String getKeyUsingTablePath(String tablePath) {
     return null;
   }
 
+  public StreamDataMap getStreamDataMap(CarbonTable table) {
```
[the remark this answers is not in the document]
Comment (QiangCai) [reply core/src/main/java/org/apache/carbondata/core/datamap/DataMapStoreManager.java]: fixed

File: core/src/main/java/org/apache/carbondata/core/datamap/StreamDataMap.java (hunk 9 above)
Comment (QiangCai) [reply core/src/main/java/org/apache/carbondata/core/datamap/StreamDataMap.java]: fixed

File: core/src/main/java/org/apache/carbondata/core/datamap/StreamFile.java (hunk 10 above)
Comment (QiangCai) [reply core/src/main/java/org/apache/carbondata/core/datamap/StreamFile.java]: fixed

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java (hunk 11 above)
Comment (QiangCai) [reply hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java]: fixed

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8302/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/232/
